src/json.hpp

```diff
@@ -122,6 +135,186 @@ struct has_mapped_type
+struct is_compatible_integer_type_impl : std::false_type{};
+
+template <typename RealIntegerType, typename CompatibleNumberIntegerType>
+struct is_compatible_integer_type_impl<true, RealIntegerType, CompatibleNumberIntegerType>
+{
+ static constexpr auto value = std::is_constructible<RealIntegerType, CompatibleNumberIntegerType>::value and
+                 std::numeric_limits<CompatibleNumberIntegerType>::is_integer and
+                 std::numeric_limits<CompatibleNumberIntegerType>::is_signed;
+};
+
+template <typename RealIntegerType, typename CompatibleNumberIntegerType>
+struct is_compatible_integer_type
+{
+  static constexpr auto value = is_compatible_integer_type_impl<std::is_arithmetic<CompatibleNumberIntegerType>::value, RealIntegerType, CompatibleNumberIntegerType>::value;
+};
+
+template <bool, typename, typename>
+struct is_compatible_unsigned_integer_type_impl : std::false_type{};
+
+template <typename RealUnsignedType, typename CompatibleNumberUnsignedType>
+struct is_compatible_unsigned_integer_type_impl<true, RealUnsignedType, CompatibleNumberUnsignedType>
+{
+ static constexpr auto value = std::is_constructible<RealUnsignedType, CompatibleNumberUnsignedType>::value and
+                 std::numeric_limits<CompatibleNumberUnsignedType>::is_integer and
+                 not std::numeric_limits<CompatibleNumberUnsignedType>::is_signed;
+};
+
+template <typename RealUnsignedType, typename CompatibleNumberUnsignedType>
+struct is_compatible_unsigned_integer_type
+{
+  static constexpr auto value = is_compatible_unsigned_integer_type_impl<std::is_arithmetic<CompatibleNumberUnsignedType>::value, RealUnsignedType, CompatibleNumberUnsignedType>::value;
+};
+
+template <typename RealFloat, typename CompatibleFloat>
+struct is_compatible_float_type
+{
+  static constexpr auto value = std::is_constructible<RealFloat, CompatibleFloat>::value and
+                                std::is_floating_point<CompatibleFloat>::value;
+};
+
+template <template <typename, typename> class JSONSerializer, typename Json,
+          typename T>
+struct has_from_json
+{
+private:
+  template <typename U, typename = decltype(uncvref_t<U>::from_json(
+                            std::declval<Json>(), std::declval<T &>()))>
+  static int detect(U &&);
+
+  static void detect(...);
+
+public:
+  static constexpr bool value = std::is_integral<decltype(
+      detect(std::declval<JSONSerializer<T, void>>()))>::value;
+};
+
+template <template <typename, typename> class JSONSerializer, typename Json,
+          typename T>
+struct has_to_json
+{
+private:
+  template <typename U, typename = decltype(uncvref_t<U>::to_json(
+                            std::declval<Json &>(), std::declval<T>()))>
+  static int detect(U &&);
+
+  static void detect(...);
+
+public:
+  static constexpr bool value = std::is_integral<decltype(
+      detect(std::declval<JSONSerializer<T, void>>()))>::value;
+};
+
+void to_json();
+void from_json();
```

Comment:
While reading all the `has_xxxx` classes I was thinking why one cannot just use expression SFINAE for these things, and actually wondered if it might had something to do with MSVC.